cql3/statements/select_statement.cc

```diff
@@ -1499,70 +1502,132 @@ void select_statement::verify_ordering_is_allowed(const restrictions::statement_
+    if (contains_alias(column)) {
+        throw exceptions::invalid_request_exception(format("Aliases are not allowed in order by clause ('{}')", column));
     }
-    for (auto&& def : selection.get_columns()) {
-        if (!def->is_partition_key() && !def->is_static()) {
-            throw exceptions::invalid_request_exception(format("SELECT DISTINCT queries must only request partition key columns and/or static columns (not {})",
-                def->name_as_text()));
+    throw exceptions::invalid_request_exception(format("Order by on unknown column {}", column));
+}
+
+select_statement::prepared_orderings_type select_statement::prepare_orderings(const schema& schema) const {
+    prepared_orderings_type prepared_orderings;
+    prepared_orderings.reserve(_parameters->orderings().size());
+
+    for (auto&& [column_id, column_ordering] : _parameters->orderings()) {
+        ::shared_ptr<column_identifier> column = column_id->prepare_column_identifier(schema);
+
+        const column_definition* def = schema.get_column_definition(column->name());
+        if (def == nullptr) {
+            handle_unrecognized_ordering_column(*column);
+        }
+
+        if (!def->is_clustering_key()) {
+            throw exceptions::invalid_request_exception(format("Order by is currently only supported on the clustered columns of the PRIMARY KEY, got {}", *column));
         }
+
+        prepared_orderings.emplace_back(def, column_ordering);
     }
 
-    // If it's a key range, we require that all partition key columns are selected so we don't have to bother
-    // with post-query grouping.
-    if (!restrictions.is_key_range()) {
+    // Uncomment this to allow specifing ORDER BY columns in any order.
+    // Right now specifying ORDER BY (c2 asc, c1 asc) is illegal, it can only be ORDER BY (c1 asc, c2 asc).
+    //
+    // std::sort(prepared_orderings.begin(), prepared_orderings.end(),
+    //     [](const std::pair<const column_definition*, ordering>& a, const std::pair<const column_definition*, ordering>& b) {
+    //         return a.first->component_index() < b.first->component_index();
+    //     }
+    // );
+
+    return prepared_orderings;
+}
+
+// Checks whether this ordering causes select results on this column to be reversed.
+// A clustering column can be ordered in the descending order in the table.
+// Then specifying ascending order would cause the results of this column to be reverse in comparison to a standard select.
+static bool are_column_select_results_reversed(const column_definition& column, select_statement::ordering column_ordering) {
+    if (column_ordering == select_statement::ordering::ascending) {
+        if (column.type->is_reversed()) {
+            return true;
+        } else {
+            return false;
+        }
+    } else { // descending ordering
+        if (column.type->is_reversed()) {
+            return false;
+        } else {
+            return true;
+        }
+    }
```

Comment:
You can replace all twelve lines of this function with one line:
```c
return (column_ordering == select_statement::ordering::descending) == column.type->is_reversed()
```

Whether it is a good idea or not, I don't know :-)

Reply:
I spent a moment trying to understand existing code which had stuff like this but I had a bad experience with that.
Each time I thought I knew what it's doing it turned out that it does the opposite ;-;
I feel this is as readable as I could make it.

Reply:
Ok. I guess it's just a matter of style.